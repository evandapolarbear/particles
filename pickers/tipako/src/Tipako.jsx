import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import baseStyles from './Tipako.scss';
import composeStyles from '../../../shared/stylesheetComposer';
import generateId from '../../../shared/generateId';

let counter = 0;
let groupCounter = 0;

export default class Tipako extends React.Component {
  static propTypes = {
    slotBottom: PropTypes.node,
    closeOnSelect: PropTypes.bool,
    data: PropTypes.arrayOf(PropTypes.shape({
      children: PropTypes.arrayOf(PropTypes.shape({
        disabled: PropTypes.bool,
        key: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
        value: PropTypes.string.isRequired
      })),
      disabled: PropTypes.bool,
      key: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
      value: PropTypes.string.isRequired
    })),
    keyField: PropTypes.string,
    loading: PropTypes.bool,
    onClear: PropTypes.func,
    onClearAll: PropTypes.func,
    onFocus: PropTypes.func,
    onSearch: PropTypes.func,
    onSelect: PropTypes.func.isRequired,
    onSelectAll: PropTypes.func,
    renderEmpty: PropTypes.func,
    renderGroup: PropTypes.func,
    renderItem: PropTypes.func,
    searchable: PropTypes.bool,
    stylesheets: PropTypes.arrayOf(PropTypes.shape()),
    titlePlaceholder: PropTypes.string,
    slotTitle: PropTypes.element,
    titleValue: PropTypes.string,
    updateOnSelect: PropTypes.bool,
    valueField: PropTypes.string,
    valueFunction: PropTypes.func
  }

  static defaultProps = {
    slotBottom: null,
    closeOnSelect: false,
    data: [],
    keyField: 'key',
    loading: false,
    onClear: null,
    onClearAll: null,
    onFocus: null,
    onSearch: null,
    onSelectAll: null,
    renderEmpty: null,
    renderGroup: null,
    renderItem: null,
    searchable: false,
    selectedKey: null,
    stylesheets: [],
    titlePlaceholder: 'Select...',
    slotTitle: null,
    titleValue: '',
    updateOnSelect: false,
    valueField: 'value',
    valueFunction: null
  }

  constructor(props) {
    super(props);

    this.styles = composeStyles(baseStyles, [...props.stylesheets]);

    this.guid = generateId();

    if (props.searchable === false && props.onSearch !== null) {
      console.error('An instance of Tipako has an "onSearch()" ' // eslint-disable-line
        + 'callback defined, but its "searchable" prop is false, '
        + 'so the callback will have no effect.');
    }

    this.state = {
      currentIndex: -1,
      expanded: false,
      value: props.titleValue
    };
  }

  componentWillMount() {
    window.addEventListener('click', this.onBlur);
  }

  componentDidUpdate() {
    const { currentIndex, expanded } = this.state;

    if (expanded) {
      if (this.focusedItem && currentIndex > -1) {
        this.focusedItem.focus();
      }

      if (this.searchInput && currentIndex === -1) {
        this.searchInput.focus();
      }
    }
  }

  componentWillUnmount() {
    window.removeEventListener('click', this.onBlur);
  }

  onSearch = (evt) => {
    const str = evt.target.value;
    this.setState({ value: str, expanded: true }, () => {
      if (this.props.onSearch) {
        this.props.onSearch(str);
      }
    });
  }

  onChildClick = (evt, child) => {
    evt.stopPropagation();

    if (child.disabled) {
      return;
    }

    this.props.onSelect(child);

    if (this.props.closeOnSelect) {
      this.setState({ expanded: false });
    }

    this.updateValue(child);
  }

  onGroupClick = (evt, group) => {
    evt.stopPropagation();

    if (group.disabled) {
      return;
    }

    this.props.onSelect(group);

    if (this.props.closeOnSelect) {
      this.setState({ expanded: false });
    }

    this.updateValue(group);
  }

  onUngroupedClick = (evt, item) => {
    evt.stopPropagation();

    if (item.disabled) {
      return;
    }

    this.props.onSelect(item);

    if (this.props.closeOnSelect) {
      this.setState({ expanded: false });
    }

    this.updateValue(item);
  }

  onCaretClick = () => {
    if (this.state.expanded === false) {
      this.setState({ currentIndex: -1, expanded: true, guid: this.guid });
      this.props.onFocus && this.props.onFocus();
    }
  }

  onSelectAll = () => {
    this.props.onSelectAll();
    this.setState({ expanded: false });
  }

  onClearAll = () => {
    this.props.onClearAll([]);
    this.setState({ expanded: false });
  }

  onSearchFocus = (evt) => {
    evt.target.select();
    this.props.onFocus && this.props.onFocus();
  }

  onBlur = () => {
    const expanded = (this.state.guid === this.guid);
    this.setState({ expanded, guid: null });
  };

  onInputBlur = () => {
    if (this.state.value.length === 0 && this.props.titleValue) {
      this.setState({ value: this.props.titleValue });
    }
  };

  onInputClear = (evt) => {
    evt.stopPropagation();

    if (this.props.onSearch) {
      this.props.onSearch('');
    }

    if (this.props.onClear) {
      this.props.onClear();
    }

    this.setState({ value: '' });
  }

  getEmptyString = () => {
    if (this.state.value) {
      return `No matches for "${this.state.value}".`;
    }

    if (this.props.loading) {
      return 'Retrieving items...';
    }

    return 'No items found.';
  }

  updateValue = (item) => {
    if (this.props.updateOnSelect) {
      if (this.props.valueFunction) {
        this.setState({ value: this.props.valueFunction(item) });
      } else {
        this.setState({ value: item[this.props.valueField] });
      }
    }
  }

  arrowKeyListener = (evt) => {
    let { currentIndex } = this.state;
    const { searchable } = this.props;

    let totalCounter = counter > groupCounter
      ? counter
      : groupCounter;

    if (groupCounter === 0) {
      totalCounter -= 1;
    }

    if (evt.keyCode === 38 && this.state.expanded) {
      evt.preventDefault();
      if ((currentIndex > -1 && searchable) || (currentIndex > 0 && !searchable)) {
        currentIndex -= 1;
        this.setState({ currentIndex });
      }
    }

    if (evt.keyCode === 40 && this.state.expanded) {
      evt.preventDefault();
      if (currentIndex < totalCounter) {
        currentIndex += 1;
        this.setState({ currentIndex });
      }
    }
  }

  render() {
    const {
      slotBottom,
      data,
      keyField,
      loading,
      onClearAll,
      onSearch,
      onSelectAll,
      renderEmpty,
      renderGroup,
      renderItem,
      searchable,
      titlePlaceholder,
      slotTitle,
      titleValue,
      updateOnSelect,
      valueField
    } = this.props;

    const { value } = this.state;

    const searchTerm = (searchable && value && !onSearch)
      ? value.toLowerCase()
      : '';

    const { currentIndex } = this.state;

    counter = 0;
    groupCounter = 0;

    const items = data.reduce((acc, v, i) => {
      // Grouped
      if (v.children) {
        const children = v.children.reduce((result, vv, ii) => {
          if (vv[valueField].toLowerCase().indexOf(searchTerm) === -1) {
            return result;
          }

          groupCounter > counter
            ? counter = groupCounter + 1
            : counter += 1;

          const focusedItem = currentIndex === counter;

          const childItem = result.concat(<div
            className={cx(this.styles.item, this.styles.childItem,
              { [this.styles.disabled]: vv.disabled,
                [this.styles.keyFocus]: focusedItem })}
            key={`child-${v[keyField]}-${vv[keyField]}`}
            onClick={(e) => { this.onChildClick(e, vv); }}
            onKeyDown={(e) => { if (focusedItem && e.keyCode === 13) this.onChildClick(e, vv); }}
            ref={(el) => { if (focusedItem) this.focusedItem = el; }}
            tabIndex={-1}
          >
            {renderItem ? renderItem(vv, ii) : vv[valueField]}
          </div>);

          return childItem;
        }, []);

        if (children.length === 0 && v[valueField].toLowerCase().indexOf(searchTerm) === -1) {
          return acc;
        }

        const focusedItem = currentIndex === groupCounter;

        const group = (<div
          className={cx(this.styles.item, this.styles.groupItem,
            { [this.styles.disabled]: v.disabled,
              [this.styles.keyFocus]: focusedItem })}
          key={`group-${v[keyField]}`}
          onClick={(evt) => { this.onGroupClick(evt, v); }}
          onKeyDown={(e) => { if (focusedItem && e.keyCode === 13) this.onGroupClick(e, v); }}
          ref={(el) => { if (focusedItem) this.focusedItem = el; }}
          tabIndex={-1}
        >
          {renderGroup ? renderGroup(v, i) : v[valueField]}
        </div>);

        groupCounter = counter + 1;
        return acc.concat(group).concat(children);
      }

      if (v[valueField].toLowerCase().indexOf(searchTerm) === -1) {
        return acc;
      }

      // Ungrouped
      if (groupCounter > 0) {
        groupCounter > counter
          ? counter = groupCounter + 1
          : counter += 1;
      }

      const focusedItem = currentIndex === counter;

      const ungrouped = (<div
        className={cx(this.styles.item, this.styles.ungroupedItem,
          { [this.styles.disabled]: v.disabled,
            [this.styles.keyFocus]: focusedItem })}
        key={`ungrouped-${v[keyField]}`}
        onClick={(evt) => { this.onUngroupedClick(evt, v); }}
        onKeyDown={(e) => { if (focusedItem && e.keyCode === 13) this.onUngroupedClick(e, v); }}
        ref={(el) => { if (focusedItem) this.focusedItem = el; }}
        tabIndex={-1}
      >
        {renderItem ? renderItem(v, i) : v[valueField]}
      </div>);

      if (groupCounter === 0) {
        counter += 1;
      }

      return acc.concat(ungrouped);
    }, []);

    const selectAll = (onSelectAll && items.length > 0)
      ? (<button className={this.styles.controlsButton} onClick={this.onSelectAll}>
           Select All
        </button>)
      : null;

    const clearAll = onClearAll
      ? (<button className={this.styles.controlsButton} onClick={this.onClearAll}>
          Clear All
        </button>)
      : null;

    const spacer = (clearAll && selectAll)
      ? <div className={this.styles.controlsSpacer}>/</div>
      : null;

    const controls =
      (selectAll || clearAll) && (
        <div className={this.styles.controls}>
          {selectAll}
          {spacer}
          {clearAll}
        </div>
      );

    const empty = (<div className={this.styles.empty}>
      {renderEmpty ? renderEmpty() : this.getEmptyString()}
    </div>);

    const caret = loading
      ? null
      : (<button onClick={this.onCaretClick} className={this.styles.caret}>
        <span className={cx('fa', 'fa-caret-down', this.styles.arrow, { [this.styles.expanded]: this.state.expanded })} />
      </button>);

    const clear = value
      ? <button onClick={this.onInputClear} className={this.styles.clear} />
      : null;

    const spinner = loading
      ? <span className={this.styles.spinner} />
      : null;

    const slot = slotTitle
      ? <div className={this.styles.slot}>{slotTitle}</div>
      : null;

    const search = searchable
      ? (
        <input
          className={cx(this.styles.input, { [this.styles.noClear]: !updateOnSelect })}
          onBlur={this.onInputBlur}
          onChange={this.onSearch}
          onClick={this.onCaretClick}
          onFocus={this.onSearchFocus}
          placeholder={titlePlaceholder}
          ref={(input) => { this.searchInput = input; }}
          type='text'
          value={value || ''}
        />
        )
      : (
        <div
          className={cx(this.styles.staticText, { [this.styles.noClear]: !updateOnSelect })}
          onClick={this.onCaretClick}
        >
          {(value.length > 0 && value) || titleValue || titlePlaceholder}
        </div>
      );

    return (
      <div
        className={cx(this.styles.container, { [this.styles.active]: this.state.expanded })}
        onKeyDown={this.arrowKeyListener}
        tabIndex={-1}
      >
        <div
          className={cx(this.styles.title, { [this.styles.expanded]: this.state.expanded })}
        >
          {slot}
          {search}
          {clear}
          {caret}
          {spinner}
        </div>

        <div className={this.styles.dropdownContainer}>
          <div
            className={cx(this.styles.dropdown, {
              [this.styles.expanded]: this.state.expanded,
              [this.styles.withSlotBottom]: slotBottom && this.state.expanded
            })}
          >
            {controls}
            <div className={this.styles.itemsContainer}>
              {items.length ? items : empty}
            </div>
            {slotBottom &&
              <div
                className={this.styles.slotBottom}
                onClick={e => e.stopPropagation()}
              >{slotBottom}</div>}
          </div>
        </div>
      </div>
    );
  }
}
