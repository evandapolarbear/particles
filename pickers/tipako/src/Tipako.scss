$h: 30px;
$p: 5px;

.active {
  background: inherit;
}

.container {
  position: relative;

  * {
    box-sizing: border-box;
  }
}

.title {
  align-items: center;
  border: 1px solid gray;
  display: flex;
  overflow: hidden;
  padding-right: $h;
  position: relative;

  .slot {
    font-family: inherit;
  }

  .staticText {
    flex-grow: 1;
    height: $h;
    line-height: $h;
    padding: 0 4px;
  }
}

.input {
  border: 0;
  cursor: text;
  display: block;
  flex-grow: 1;
  font-size: 13px;
  height: $h;
  line-height: $h;
  padding: 0 24px 0 4px;
}

.inputIcon {
  background: 0;
  border: 0;
  display: block;
  font-size: 12px;
  height: $h;
  line-height: $h;
  position: absolute;
  right: 0;
  text-align: center;
  top: 0;
  width: $h;

  &:hover {
    background: #ddd;
    cursor: pointer;
  }
}

.caret {
  composes: inputIcon;
  border-left: 1px solid #999;
  cursor: pointer;
  display: block;

  .arrow {
    cursor: pointer;
    display: block;
    font-size: 13px;
    margin-top: 2px;
    transition: transform 0.3s ease;

    &.expanded {
      transform: rotate(180deg);
    }
  }

  &:focus {
    outline: 0;
  }
}

.clear {
  composes: inputIcon;
  right: $h;

  &::before {
    content: '\00D7';
    display: block;
    font-size: 16px;
    height: 16px;
    line-height: 16px;
  }
}

.spinner {
  composes: inputIcon;
  animation-duration: 1.5s;
  animation-name: spin;
  animation-iteration-count: infinite;
  animation-timing-function: linear;
  background-image:
    url(
      "data:image/svg+xml;base64,PHN2ZyB4bWxucz\
      0naHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmcnIHdpZHRoPScxNicgaGVpZ2h\
      0PScxNic+PGVsbGlwc2UgY3g9IjgiIGN5PSI4IiByeD0iNyIgcnk9IjciIHN0\
      cm9rZS13aWR0aD0iMiIgc3Ryb2tlLW9wYWNpdHk9JzAuNScgc3Ryb2tlPSIjY\
      2NjIiBmaWxsPSJub25lIiAvPjxwYXRoIGQ9J00gMSA4IEEgOCA4IDAgMCAwID\
      QgMTMuNycgZmlsbD0nbm9uZScgc3Ryb2tlPScjRjQ3RTQyJyAgc3Ryb2tlLXd\
      pZHRoPScyJyBzdHJva2Utb3BhY2l0eT0nMC44JyAvPjwvc3ZnPg=="
    );
  background-position: center center;
  background-repeat: no-repeat;
  background-size: 16px 16px;
}

.dropdownContainer {
  position: relative;
}

.dropdown {
  border: 0;
  box-sizing: border-box;
  cursor: pointer;
  max-height: 0;
  margin-top: -1px;
  overflow: hidden;
  position: absolute;
  transition: all 0.15s ease;
  width: 100%;

  &.expanded {
    background: #fff;
    border: 1px solid #666;
    height: auto;
    max-height: 300px;
    z-index: 100;
  }
}

.empty {
  padding: 10px;
}

.item {
  align-items: center;
  background: 0;
  border: 0;
  cursor: pointer;
  display: flex;
  flex-direction: row;
  margin-bottom: 3px;
  text-align: left;
  padding: 2px $p;
  width: 100%;

  * {
    cursor: pointer;
  }

  &.disabled,
  &.disabled *, {
    color: #aaa;
    cursor: not-allowed;
  }
}

.itemIcon {
  height: $h;
  margin-right: 5px;
  display: flex;
  width: $h;
}

.itemText {
}

.groupItem {
  // color: seagreen;
}

.childItem {
  // color: firebrick;
  padding: 2px ($p + 10);
}

.ungroupedItem {
  // color: turquoise;
  cursor: pointer;
}

.controls {
  line-height: $h;
  margin: 5px 0;
  padding: 0 10px;

  .controlsButton {
    background: 0;
    border: 0;
    border-radius: 2px;
    cursor: pointer;
    display: inline-block;
    font-size: 13px;
    padding: 3px;
    text-decoration: none;
    vertical-align: middle;

    &:hover {
      text-decoration: underline;
    }
  }

  .controlsSpacer {
    display: inline-block;
    margin: 0 3px;
    vertical-align: middle;
  }

  &:hover {
    background: #e7e7e7;
  }
}

.itemsContainer {
  max-height: 250px;
  overflow-y: auto;
}

@keyframes spin {
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
}
