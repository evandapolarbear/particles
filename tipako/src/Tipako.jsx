import React, { PropTypes } from 'react';
import cx from 'classnames';
import styles from './Tipako.scss';
import procoreStyles from './TipakoProcore.scss';
import procoreOverrides from './TipakoProcoreOverrides.scss';

let timer = null;

export default class Tipako extends React.Component {
  static propTypes = {
    addGroupTokens: PropTypes.bool,
    className: PropTypes.string,
    data: PropTypes.arrayOf(PropTypes.shape({
      children: PropTypes.arrayOf(PropTypes.shape({
        text: PropTypes.string.isRequired,
      })),
      text: PropTypes.string.isRequired,
      id: PropTypes.number.isRequired,
    })),
    groupIcon: PropTypes.string,
    itemIcon: PropTypes.string,
    maxResults: PropTypes.number,
    msgEmpty: PropTypes.string,
    msgPlaceholder: PropTypes.string,
    onFetch: PropTypes.func,
    onSelect: PropTypes.func.isRequired,
    renderTokens: PropTypes.func,
    stylesheets: PropTypes.arrayOf(PropTypes.shape()),
  }

  static defaultProps = {
    addGroupTokens: false,
    className: '',
    data: [],
    groupIcon: null,
    itemIcon: null,
    maxResults: Math.Infinite,
    msgEmpty: 'No matching items.',
    msgPlaceholder: 'Search...',
    onFetch: () => {},
    renderTokens: null,
    stylesheets: [],
  }

  constructor(props) {
    super(props);

    // TEMPORARY...? Ben 170217
    props.stylesheets.push(procoreStyles);
    props.stylesheets.push(procoreOverrides);

    props.stylesheets.forEach((s) => {
      Object.keys(s).forEach((k) => {
        if (styles[k] && s[k]) {
          styles[k] = styles[k].split(' ').concat(s[k].split(' ')).join(' ');
        }
      });
    });

    this.state = {
      data: this.props.data || [],
      expanded: false,
      fetching: false,
      tokens: {},
      value: '',
    };
  }

  componentWillMount() {
    window.addEventListener('click', this.onBlur);
  }

  componentDidMount() {
    this.props.onFetch('', (data) => {
      this.setState({ data, fetching: false });
    });
  }

  onSearch = (evt) => {
    const str = evt.target.value;

    if (this.props.onFetch === undefined) {
      const data = this.props.data.reduce((acc, val) => {
        const matchText = val.text.toLowerCase().indexOf(str) !== -1;

        const matchChild = val.children ||
          val.children.reduce((result, child) => {
            return result || (child.text.toLowerCase().indexOf(str) !== -1);
          }, false);

        if (matchText || matchChild) {
          return acc.concat(val);
        }

        return acc;
      }, []);

      this.setState({ data, expanded: true });
    } else {
      this.setState({ value: str, fetching: true });
      clearTimeout(timer);

      timer = setTimeout(() => {
        this.props.onFetch(str, (data) => {
          this.setState({ data, expanded: true, fetching: false });
        });
      }, 500);
    }

    this.setState({ value: str });
  }

  onChildClick = (obj) => {
    const tokens = this.state.tokens;
    tokens[obj.id] = obj;

    this.setState({ tokens, expanded: false, value: '' });
  }

  onGroupClick = (obj) => {
    const tokens = this.state.tokens;

    if (this.props.addGroupTokens) {
      tokens[obj.id] = obj;
    }

    obj.children.forEach((child) => {
      tokens[child.id] = child;
    });

    this.props.onSelect(Object.values(tokens));
    this.setState({ tokens, expanded: false, value: '' });
  }

  onUngroupedClick = (obj) => {
    const tokens = this.state.tokens;
    tokens[obj.id] = obj;

    this.props.onSelect(Object.values(tokens));
    this.setState({ tokens, expanded: false, value: '' });
  }

  onTokenClick = (obj) => {
    const tokens = this.state.tokens;
    delete tokens[obj.id];

    this.props.onSelect(Object.values(tokens));
    this.setState({ tokens, expanded: false });
  }

  onCaretClick = (evt) => {
    evt.stopPropagation();
    this.setState({ expanded: true });
  }

  onSelectAll = () => {
    const { addGroupTokens } = this.props;

    const tokens = this.state.data.reduce((acc, obj) => {
      const result = acc;

      if (obj.children) {
        if (addGroupTokens) {
          result[obj.id] = obj;
        }

        obj.children.forEach((child) => {
          result[child.id] = child;
        });
      } else {
        result[obj.id] = obj;
      }

      return result;
    }, {});

    this.props.onSelect(Object.values(tokens));
    this.setState({ tokens, expanded: false });
  }

  onClearAll = () => {
    this.props.onSelect(Object.values(this.state.tokens));
    this.setState({ tokens: {}, expanded: false });
  }

  onFocus = () => {
    this.setState({ expanded: true });
  }

  onBlur = () => {
    this.setState({ expanded: false });
  };

  render() {
    const itemIcon = this.props.itemIcon
      ? <img alt="Item" src={this.props.itemIcon} className={styles.itemIcon} />
      : null;

    const groupIcon = this.props.groupIcon
      ? <img alt="Group" src={this.props.groupIcon} className={styles.itemIcon} />
      : null;

    const items = this.state.data.reduce((acc, v) => {
      // Grouped
      if (v.children) {
        const children = v.children.reduce((result, vv) => {
          if (this.state.tokens[vv.id] !== undefined) {
            return result;
          }

          return result.concat(<button
            onClick={() => { this.onChildClick(vv); }}
            className={`${styles.item} ${styles.childItem}`}
            key={`item-${vv.id}`}
          >
            {itemIcon}
            {vv.text}
          </button>);
        }, []);

        if (this.props.addGroupTokens === false && children.length === 0) {
          return acc;
        }

        if (this.state.tokens[v.id] !== undefined) {
          return acc;
        }

        const group = (<button
          onClick={() => { this.onGroupClick(v); }}
          className={`${styles.item} ${styles.groupItem}`}
          key={`item-${v.id}`}
        >
          {groupIcon}
          {v.text}
        </button>);

        return acc.concat(group).concat(children);
      }

      // Ungrouped
      if (this.state.tokens[v.id] !== undefined) {
        return acc;
      }

      const ungrouped = (<button
        onClick={() => { this.onUngroupedClick(v); }}
        className={`${styles.item} ${styles.ungroupedItem}`}
        key={`item-${v.id}`}
      >
        {itemIcon}
        {v.text}
      </button>);

      return acc.concat(ungrouped);
    }, []);

    const selectall = (<div className={styles.controls}>
      <button
        className={styles.controlsButton}
        onClick={this.onSelectAll}
      >
         Select All
      </button>
      <div className={styles.controlsSpacer}>/</div>
      <button
        className={styles.controlsButton}
        onClick={this.onClearAll}
      >
        Clear All
      </button>
    </div>);

    const values = Object.values(this.state.tokens);
    const tokens = this.props.renderTokens
       ? this.props.renderTokens(values, this.onTokenClick)
       : values.map((val) => {
         return (<button
           key={`token-${val.id}`}
           onClick={() => { this.onTokenClick(val); }}
         >
           {val.text}
         </button>);
       });

    const nomatch = <div className={styles.nomatch}>{this.props.msgEmpty}</div>;

    const caret = (this.state.fetching || items.length === 0)
      ? null
      : (<button
        className={`fa fa-caret-down ${styles.caret}`}
        onClick={this.onCaretClick}
      />);

    const busy = this.state.fetching
      ? <span className={styles.busy} />
      : null;

    const maxResultsWarningIcon =
      (this.state.fetching || this.state.data.length < this.props.maxResults)
      ? null
      : <span className={`fa fa-exclamation-triangle ${styles.maxResults}`} />;

    const maxResultsWarningText =
      (this.state.fetching || items.length < this.props.maxResults)
      ? ''
      : 'This search is too general, so the results have been limited.';

    return (<div
      className={cx(styles.picker, this.props.className)}
      title={maxResultsWarningText}
    >
      {tokens}

      <div className={cx(styles.inputContainer)}>
        <input
          className={styles.input}
          type="text"
          placeholder={this.props.msgPlaceholder}
          value={this.state.value}
          onChange={this.onSearch}
        />
        {caret}
        {busy}
        {maxResultsWarningIcon}
      </div>

      <div className={styles.dropdownContainer}>
        <div
          className={cx(styles.dropdown, {
            [styles.expanded]: this.state.expanded })}
        >
          {items.length > 0 ? selectall : ''}
          {items.length ? items : nomatch}
        </div>
      </div>
    </div>);
  }
}
