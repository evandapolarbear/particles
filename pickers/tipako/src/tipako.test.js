import React from 'react';
import Tipako from './Tipako.jsx';
import props from '../sandbox/data.js';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(
    <Tipako {...props} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
