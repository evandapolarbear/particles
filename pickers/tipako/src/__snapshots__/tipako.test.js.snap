// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className=""
  onKeyDown={[Function]}
  tabIndex={-1}
>
  <div
    className=""
  >
    <div
      className={undefined}
    >
      <div
        className="fa fa-bus undefined"
      />
    </div>
    <input
      className="undefined"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      placeholder="Make a selection..."
      type="text"
      value="Calypso"
    />
    <button
      className={undefined}
      onClick={[Function]}
      type="button"
    />
    <button
      className={undefined}
      onClick={[Function]}
      type="button"
    >
      <span
        className="fa fa-caret-down"
      />
    </button>
  </div>
  <div
    className={undefined}
  >
    <div
      className=""
    >
      <div
        className={undefined}
      >
        <button
          className={undefined}
          onClick={[Function]}
          type="button"
        >
          Select All
        </button>
        <div
          className={undefined}
        >
          /
        </div>
        <button
          className={undefined}
          onClick={[Function]}
          type="button"
        >
          Clear All
        </button>
      </div>
      <div
        className={undefined}
      >
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Animals
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Dolphin
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Albatross
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          German Shepherd
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Mole
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Lesser Southern Floridian Muskrat
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Woodworking
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Lutherie
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Carpentry
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Parquetry
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Joinery
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Cabinetry
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Saturnian Moons Found Around Saturn
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Tethys
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Calypso
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Enceladus
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Rhea
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Iapetus
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          An empty group appears!
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Magnetic poles are points to and from which the lines of magnetic force are directed.
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Guacamole
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Crankshaft
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Demographic
        </div>
        <div
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Almond
        </div>
      </div>
      <div
        className={undefined}
        onClick={[Function]}
      >
        <div
          className={undefined}
        >
          Bottom Slot
        </div>
      </div>
    </div>
  </div>
</div>
`;
